### PR TITLE
[618] Migration from ViewPager to ViewPager2 

### PR DESCRIPTION
**Pull request by arangob**
Fixes issue #618 - Update the About screen to use the ViewPager2 library

### Issues encountered & feedback
Overall smooth migration. However new ViewPager2 does not have pageMargin variable, so I couldn't figure out how to keep the space between pages.

*This is my first contribution to an open source Android project. Feedback welcomed!

### Steps taken
1. Added dependency
2. Switched from `ViewPager` to `ViewPager2` in `activity_about.xml`
3. Modified `AboutActivity.java` to use new view pager
4. Replaced old `PagerAdapter` extension with `RecyclerView.Adapter` in `AboutPagerAdapter.java` and implemented member functions
5. Removed old `ViewPager.OnPageChangeListener` implementation in `InkPageIndicator.java` and replaced it with new `registerOnPageChangeCallback`

### CODE REVIEW

File: about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt [hunk 1]
```diff
@@ -30,33 +29,36 @@ import java.security.InvalidParameterException
...
 
     private var aboutPlaid: View? = null
     private var aboutIcon: View? = null
     private var aboutLibs: View? = null
 
-    override fun instantiateItem(collection: ViewGroup, position: Int): Any {
-        return getPage(position, collection).also {
-            collection.addView(it)
+    override fun getItemViewType(position: Int): Int {
+        return when (position) {
```
Comment (florina-muntenescu) [inline about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt]: use `when(...){...}.exhaustive`

File: about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt [hunk 2]
```diff
@@ -30,33 +29,36 @@ import java.security.InvalidParameterException
...
+            2 -> 2
+            else -> throw InvalidParameterException()
         }
     }
 
-    override fun destroyItem(collection: ViewGroup, position: Int, view: Any) {
-        collection.removeView(view as View)
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AboutViewHolder {
+        return when (viewType) {
+            0 -> AboutViewHolder(getAboutAppPage(parent))
```
Comment (florina-muntenescu) [inline about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt]: I liked the previous approach of handling 2 and throwing an exception for else. could we apply it here also?

File: about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt [hunk 3]
```diff
@@ -30,33 +29,36 @@ import java.security.InvalidParameterException
...
+            1 -> AboutViewHolder(getAboutIconPage(parent))
+            else -> AboutViewHolder(getAboutLibsPage(parent))
+        }
```
Comment (florina-muntenescu) [inline about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt]: `.exhaustive`

File: about/src/main/java/io/plaidapp/about/ui/widget/InkPageIndicator.java [hunk 4]
```diff
@@ -159,14 +158,54 @@ public InkPageIndicator(Context context, AttributeSet attrs, int defStyle) {
...
 
-    public void setViewPager(ViewPager viewPager) {
+    public void setViewPager(ViewPager2 viewPager) {
         this.viewPager = viewPager;
-        viewPager.addOnPageChangeListener(this);
-        setPageCount(viewPager.getAdapter().getCount());
-        viewPager.getAdapter().registerDataSetObserver(new DataSetObserver() {
+        viewPager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
+            @Override
+            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
```
Comment (florina-muntenescu) [inline about/src/main/java/io/plaidapp/about/ui/widget/InkPageIndicator.java]: several lines of this method exceed the 100 char limit, could you split them on multiple lines?

File: core_dependencies.gradle [hunk 5]
```diff
@@ -23,6 +23,7 @@ dependencies {
     implementation "androidx.palette:palette:${versions.androidx}"
     implementation "androidx.recyclerview:recyclerview:${versions.androidx}"
     implementation "androidx.dynamicanimation:dynamicanimation:${versions.androidx}"
+    implementation "androidx.viewpager2:viewpager2:${versions.viewPager2}"
```
Comment (nickbutcher) [inline core_dependencies.gradle]: As this is only used in `:about` lets add the dependency there to minimize its scope.

**Comment by florina-muntenescu** [comment]
Issue for the lack of `pageMargin` in ViewPager2 [created](https://issuetracker.google.com/127463284).
Thank you, @arangob for helping us improve Android APIs!

File: about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt (hunk 2 above)
Comment (arangob) [reply about/src/main/java/io/plaidapp/about/ui/adapter/AboutPagerAdapter.kt]: yes, added

File: about/src/main/java/io/plaidapp/about/ui/widget/InkPageIndicator.java (hunk 4 above)
Comment (arangob) [reply about/src/main/java/io/plaidapp/about/ui/widget/InkPageIndicator.java]: fixed!

File: core_dependencies.gradle (hunk 5 above)
Comment (arangob) [reply core_dependencies.gradle]: done!

**Comment by florina-muntenescu** [comment]
@arangob Looks like we have a fix for the margin:
```
binding.pager.apply {
       adapter = AboutPagerAdapter(viewModel.uiModel)

       // Set the margin between pages in the ViewPager2
       val pageMargin = resources.getDimensionPixelSize(appR.dimen.spacing_normal)
       setPageTransformer { page, position -> page.translationX = position * pageMargin }
}
```

Could you add this change and rebase on top of master? 
Then we'll be able to merge this!

**Comment by arangob** [comment]
@florina-muntenescu ok, added the change!

**Comment by florina-muntenescu** [comment]
@arangob - could you also rebase and fix the conflicts? 
Without this we can't merge your PR.

**Comment by arangob** [comment]
@florina-muntenescu I've rebased and now there is a CLA conflict. I'm not sure what the issue it.

**Comment by arangob** [comment]
@florina-muntenescu I rebase with master and now it appears there is a CLA conflict. It's not clear to me why that is. Thank you!

**Comment by florina-muntenescu** [comment]
@arangob - Looks like you merged the master into your branch, instead of rebasing. 
For rebase would have done the following: 
* [Sync your fork](https://help.github.com/en/articles/syncing-a-fork) so your master can be up to date
* In your branch call `git rebase master` 
* Fix conflicts
* Force push the rebase fixes and call `git push -f origin ba/viewPager2`

Given the current state of the branch, I would suggest:
1. Create a new branch from master
2. Cherry-pick your changes one by one from this branch onto the new branch. 
3. Create a new PR.

**Comment by arangob** [comment]
> @arangob - Looks like you merged the master into your branch, instead of rebasing.
> For rebase would have done the following:
> 
> * [Sync your fork](https://help.github.com/en/articles/syncing-a-fork) so your master can be up to date
> * In your branch call `git rebase master`
> * Fix conflicts
> * Force push the rebase fixes and call `git push -f origin ba/viewPager2`
> 
> Given the current state of the branch, I would suggest:
> 
> 1. Create a new branch from master
> 2. Cherry-pick your changes one by one from this branch onto the new branch.
> 3. Create a new PR.

@florina-muntenescu yikes! i will follow the above steps next time. thank you!

OK closing this branch. New PR coming up.

**Pull request closed without merging**